File: crates/dashboard-model/src/victron_controller/dashboard/soc_projection_segment.rs
use crate::victron_controller::dashboard::soc_projection_kind::SocProjectionKind;

#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
pub struct SocProjectionSegment {
    #[serde(deserialize_with = "crate::baboon_runtime::lenient_numeric::deserialize")]
    pub start_epoch_ms: i64,
    #[serde(deserialize_with = "crate::baboon_runtime::lenient_numeric::deserialize")]
    pub end_epoch_ms: i64,
    pub start_soc_pct: f64,
    pub end_soc_pct: f64,
    pub kind: SocProjectionKind,
}

impl PartialEq for SocProjectionSegment {
    fn eq(&self, other: &Self) -> bool {
        self.cmp(other) == std::cmp::Ordering::Equal
    }
}

impl Eq for SocProjectionSegment {}

impl PartialOrd for SocProjectionSegment {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for SocProjectionSegment {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        match self.start_epoch_ms.cmp(&other.start_epoch_ms) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        match self.end_epoch_ms.cmp(&other.end_epoch_ms) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        match self.start_soc_pct.total_cmp(&other.start_soc_pct) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        match self.end_soc_pct.total_cmp(&other.end_soc_pct) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        match self.kind.cmp(&other.kind) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        std::cmp::Ordering::Equal
    }
}

impl crate::baboon_runtime::BaboonBinCodecIndexed for SocProjectionSegment {
    fn index_elements_count(_ctx: &crate::baboon_runtime::BaboonCodecContext) -> u16 {
        0
    }
}

impl crate::baboon_runtime::BaboonBinEncode for SocProjectionSegment {
    fn encode_ueba(&self, ctx: &crate::baboon_runtime::BaboonCodecContext, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        let value = self;
        if ctx.use_indices() {
            crate::baboon_runtime::bin_tools::write_byte(writer, 0x01)?;
            let mut buffer: Vec<u8> = Vec::new();
            value.start_epoch_ms.encode_ueba(ctx, &mut buffer)?;
            value.end_epoch_ms.encode_ueba(ctx, &mut buffer)?;
            value.start_soc_pct.encode_ueba(ctx, &mut buffer)?;
            value.end_soc_pct.encode_ueba(ctx, &mut buffer)?;
            value.kind.encode_ueba(ctx, &mut buffer)?;
            writer.write_all(&buffer)?;
        } else {
            crate::baboon_runtime::bin_tools::write_byte(writer, 0x00)?;
            value.start_epoch_ms.encode_ueba(ctx, writer)?;
            value.end_epoch_ms.encode_ueba(ctx, writer)?;
            value.start_soc_pct.encode_ueba(ctx, writer)?;
            value.end_soc_pct.encode_ueba(ctx, writer)?;
            value.kind.encode_ueba(ctx, writer)?;
        }
        Ok(())
    }
}

impl crate::baboon_runtime::BaboonBinDecode for SocProjectionSegment {
    fn decode_ueba(ctx: &crate::baboon_runtime::BaboonCodecContext, reader: &mut dyn std::io::Read) -> Result<Self, Box<dyn std::error::Error>> {
        let (_header, index) = <Self as crate::baboon_runtime::BaboonBinCodecIndexed>::read_index(ctx, reader)?;
        if ctx.use_indices() {
            assert_eq!(index.len(), <Self as crate::baboon_runtime::BaboonBinCodecIndexed>::index_elements_count(ctx) as usize);
        }
        let start_epoch_ms = crate::baboon_runtime::bin_tools::read_i64(reader)?;
        let end_epoch_ms = crate::baboon_runtime::bin_tools::read_i64(reader)?;
        let start_soc_pct = crate::baboon_runtime::bin_tools::read_f64(reader)?;
        let end_soc_pct = crate::baboon_runtime::bin_tools::read_f64(reader)?;
        let kind = SocProjectionKind::decode_ueba(ctx, reader)?;
        Ok(SocProjectionSegment {
            start_epoch_ms,
            end_epoch_ms,
            start_soc_pct,
            end_soc_pct,
            kind,
        })
    }
}

File: web/src/model/victron_controller/dashboard/SocProjectionSegment.ts
// @ts-nocheck
import {BaboonGeneratedLatest, BaboonCodecContext, BaboonBinWriter, BinTools, BaboonBinReader, Lazy} from '../../BaboonSharedRuntime'
import {SocProjectionKind, SocProjectionKind_UEBACodec} from './SocProjectionKind'

export class SocProjectionSegment implements BaboonGeneratedLatest {
    private readonly _start_epoch_ms: bigint;
    private readonly _end_epoch_ms: bigint;
    private readonly _start_soc_pct: number;
    private readonly _end_soc_pct: number;
    private readonly _kind: SocProjectionKind;

    constructor(start_epoch_ms: bigint, end_epoch_ms: bigint, start_soc_pct: number, end_soc_pct: number, kind: SocProjectionKind) {
        this._start_epoch_ms = start_epoch_ms
        this._end_epoch_ms = end_epoch_ms
        this._start_soc_pct = start_soc_pct
        this._end_soc_pct = end_soc_pct
        this._kind = kind
    }

    public get start_epoch_ms(): bigint {
        return this._start_epoch_ms;
    }
    public get end_epoch_ms(): bigint {
        return this._end_epoch_ms;
    }
    public get start_soc_pct(): number {
        return this._start_soc_pct;
    }
    public get end_soc_pct(): number {
        return this._end_soc_pct;
    }
    public get kind(): SocProjectionKind {
        return this._kind;
    }

    public toJSON(): Record<string, unknown> {
        return {
            start_epoch_ms: this._start_epoch_ms,
            end_epoch_ms: this._end_epoch_ms,
            start_soc_pct: this._start_soc_pct,
            end_soc_pct: this._end_soc_pct,
            kind: this._kind
        };
    }

    public with(overrides: {start_epoch_ms?: bigint; end_epoch_ms?: bigint; start_soc_pct?: number; end_soc_pct?: number; kind?: SocProjectionKind}): SocProjectionSegment {
        return new SocProjectionSegment(
            'start_epoch_ms' in overrides ? overrides.start_epoch_ms! : this._start_epoch_ms,
            'end_epoch_ms' in overrides ? overrides.end_epoch_ms! : this._end_epoch_ms,
            'start_soc_pct' in overrides ? overrides.start_soc_pct! : this._start_soc_pct,
            'end_soc_pct' in overrides ? overrides.end_soc_pct! : this._end_soc_pct,
            'kind' in overrides ? overrides.kind! : this._kind
        );
    }

    public static fromPlain(obj: {start_epoch_ms: bigint; end_epoch_ms: bigint; start_soc_pct: number; end_soc_pct: number; kind: SocProjectionKind}): SocProjectionSegment {
        return new SocProjectionSegment(
            obj.start_epoch_ms,
            obj.end_epoch_ms,
            obj.start_soc_pct,
            obj.end_soc_pct,
            obj.kind
        );
    }

    public static readonly BaboonDomainVersion = '0.3.0'
    public baboonDomainVersion() {
        return SocProjectionSegment.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return SocProjectionSegment.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#SocProjectionSegment'
    public baboonTypeIdentifier() {
        return SocProjectionSegment.BaboonTypeIdentifier
    }
    public static readonly BaboonSameInVersions = ["0.2.0", "0.3.0"]
    public baboonSameInVersions() {
        return SocProjectionSegment.BaboonSameInVersions
    }
    public static binCodec(): SocProjectionSegment_UEBACodec {
        return SocProjectionSegment_UEBACodec.instance
    }
}

export class SocProjectionSegment_UEBACodec {
    public encode(ctx: BaboonCodecContext, value: SocProjectionSegment, writer: BaboonBinWriter): unknown {
        if (this !== SocProjectionSegment_UEBACodec.lazyInstance.value) {
          return SocProjectionSegment_UEBACodec.lazyInstance.value.encode(ctx, value, writer)
        }
    
        if (ctx === BaboonCodecContext.Indexed) {
            BinTools.writeByte(writer, 0x01);
            const buffer = new BaboonBinWriter();
            BinTools.writeI64(buffer, value.start_epoch_ms);
            BinTools.writeI64(buffer, value.end_epoch_ms);
            BinTools.writeF64(buffer, value.start_soc_pct);
            BinTools.writeF64(buffer, value.end_soc_pct);
            SocProjectionKind_UEBACodec.instance.encode(ctx, value.kind, buffer);
            writer.writeAll(buffer.toBytes());
        } else {
            BinTools.writeByte(writer, 0x00)
            BinTools.writeI64(writer, value.start_epoch_ms);
            BinTools.writeI64(writer, value.end_epoch_ms);
            BinTools.writeF64(writer, value.start_soc_pct);
            BinTools.writeF64(writer, value.end_soc_pct);
            SocProjectionKind_UEBACodec.instance.encode(ctx, value.kind, writer);
        }
    }
    
    public decode(ctx: BaboonCodecContext, reader: BaboonBinReader): SocProjectionSegment {
        if (this !== SocProjectionSegment_UEBACodec .lazyInstance.value) {
            return SocProjectionSegment_UEBACodec.lazyInstance.value.decode(ctx, reader)
        }
    
        const header = BinTools.readByte(reader);
        const useIndices = header === 0x01;
        if (useIndices) {
            for (let i = 0; i < 0; i++) {
                BinTools.readI32(reader);
                BinTools.readI32(reader);
            }
        }
        const start_epoch_ms = BinTools.readI64(reader);
        const end_epoch_ms = BinTools.readI64(reader);
        const start_soc_pct = BinTools.readF64(reader);
        const end_soc_pct = BinTools.readF64(reader);
        const kind = SocProjectionKind_UEBACodec.instance.decode(ctx, reader);
        return new SocProjectionSegment(
            start_epoch_ms,
            end_epoch_ms,
            start_soc_pct,
            end_soc_pct,
            kind,
        );
    }

    public static readonly BaboonDomainVersion = '0.3.0'
    public baboonDomainVersion() {
        return SocProjectionSegment_UEBACodec.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return SocProjectionSegment_UEBACodec.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#SocProjectionSegment'
    public baboonTypeIdentifier() {
        return SocProjectionSegment_UEBACodec.BaboonTypeIdentifier
    }

    protected static lazyInstance = new Lazy(() => new SocProjectionSegment_UEBACodec())
    public static get instance(): SocProjectionSegment_UEBACodec {
        return SocProjectionSegment_UEBACodec.lazyInstance.value
    }
}